### ISSUE TITLE
Windows dynamic linking

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#2460]
Investigate dynamic linking under windows:

https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/
https://bitbucket.org/vlasovmaksim/cmake_windows_export_all_symbols/overview
https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html

**Comment by bangerth** [comment]
Ah, neat. This makes things not only simpler but feasible!

**Comment by tjhei** [comment]
I tested with a cmake 3.6 nightly and we can automatically export symbols, but I am now hitting
```The limit of 65535 objects or members in a library has been exceeded```
see https://support.microsoft.com/en-us/kb/949945. Sigh.

**Comment by bangerth** [comment]
Seriously? We're in 2016 and they can't (i) version their file formats, and (ii) offer a third or fourth byte for the number of exported members? It's hard to express how disappointed I find myself in how crappy their toolchain is :-(

**Comment by bangerth** [comment]
By the way, is this a 32-bit build? Does their 64-bit library format offer a bit more space?

**Comment by tjhei** [comment]
> By the way, is this a 32-bit build?

This is a 64bit build. The debug lib is trying to export about 300'000 symbols so splitting into something like 2d/3d wouldn't be enough.

I am somewhat frustrated, because static linking is just too slow to be practical.

**Comment by loylick** [comment]
I do not understand why you want to export ALL symbols.  Is not it enough for a library to export the interface functions and classes?

**Comment by bangerth** [comment]
That would of course be enough. But it would require annotating thousands or tens of thousands of places :-(

**Comment by tjhei** [comment]
> That would of course be enough. But it would require annotating thousands or tens of thousands of places :-(

I just had another idea. Maybe we could abuse our .inst system to generate a list of instantiations to export by generating a .def file (https://msdn.microsoft.com/en-us/library/d91k01sh.aspx). Or we bite the bullet and try ``__declspec( dllexport )`` and see how bad that is.

I wonder if we can avoid the 65k that way.

**Comment by tamiko** [comment]
> the bullet and try ``__declspec( dllexport )`` and see how bad that is.

Placement and what you need to export is incompatible with the syntax
expected by gcc/clang for *nix.

So, this will be very likely a massive windows-only code change.

With severe implications on development. We would have to test every
pull-request for correct dllexport annotations. *And* would have to
bother everyone to correctly annotate export statements.

**Comment by bangerth** [comment]
I think the only other feasible approach is to bite the bullet and make the annotations. We'd use a string such as `DEAL_II_DLL_EXPORT` that simply expands to nothing on platforms where we don't use it.

My main concern is exactly what @tamiko says in the last paragraph -- this is going to be a struggle forever. I would also expect that we will come close to the 65k limit anyway: we already declare ~800 classes, some with a lot of member functions, many to be multiplied by the number of different possible template arguments.

**Comment by tamiko** [comment]
> I think the only other feasible approach is to bite the bullet and
> make the annotations.

Personally, I think, a well maintained (manually assembled) export list
would be much cleaner than to randomly slam annotations around in
source.

What exactly speaks against an export list?


Further, I guess, we will still hit this 65k limit. I guess, it is time
to split the library again :-D

**Comment by bangerth** [comment]
Name mangling is the problem with a manual list.

**Comment by tamiko** [comment]
> Name mangling is the problem with a manual list. 

You have to update this list on Windows, of course. And there are a
number of procedures available to do this (semi)automatically.

(With CMake auto-export being the most convenient one.)

Randomly slabbing DEAL_II_DLL_EXPORT around statement on Linux (that
will never get tested) is in my opinion far worse.

**Comment by tamiko** [comment]
@tjhei The inst-files won't be enough. We would also have to maintain a separate list of all symbols that are instantiated by CC files (due to non-template code or other means).

@tjhei @bangerth How much smaller compared to just exporting all symbols would be a hand-maintained list anyway?

I see the problem that we never bothered to clearly distinguish implementation and interface in the last 10 years. On the contrary, we have the habit of exposing configuration and implementation directly to the user (for example, the same version of deal.II with different features enabled/disabled is ABI incompatible).

I fear tha, in order to be able to define a sane set of interfaces (on an API, as well as a symbol export level) we would first have to restructure and reorganize the whole library. (For example in interface headers for external use and headers for internal use). While doing so, we would have to break up the library in smaller parts as well.

Anything short of this would probably be not good enough to get dynamic linking on Windows. And this is truly a gigantic amount of work.

**Comment by tamiko** [comment]
@tjhei Btw, I remember now. Even with the cmake feature, we would need to slab DLL_IMPORT statements around all declarations. At that point I gave up in frustration.

**Comment by tjhei** [comment]
> @tjhei Btw, I remember now. Even with the cmake feature, we would need to slab DLL_IMPORT statements around all declarations. At that point I gave up in frustration.

No, only for _data_ members, right? See [WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html#prop_tgt:WINDOWS_EXPORT_ALL_SYMBOLS)

**Comment by tjhei** [comment]
So our options are:
1. no windows support
2. go with a non-msvc compiler (and also not an MSVC emulating compiler like clang/c2, but something like mingw)
3. split the library into several libraries (and use WINDOWS_EXPORT_ALL_SYMBOLS, export data members, etc.). Note that we need about 10 libraries for that to work.
4. only do static linking (I don't consider this an option)

**Comment by tamiko** [comment]
> So our options are:

> 1. no windows support

> 2. go with a non-msvc compiler (and also not an MSVC emulating compiler like clang/c2, but something like mingw)

I do not consuder (2.) to be an option either. Cygwin, MinGW, MinGW-W64
had the last time that I checked broken toolchains with ICE wrt name
mangling and maximum section support in dynamic objects.

> 3. split the library into several libraries (and use WINDOWS_EXPORT_ALL_SYMBOLS, export data members, etc.). Note that we need about 10 libraries for that to work.
> 4. only do static linking (I don't consider this an option)

5. Wait for the native Linux ABI for Windows 10 to be available for the
broad masses and go with the sources conveniently packaged for Ubuntu
(universe) [1,2,3].


[1] A nice byproduct of the Debian packaging attempts.

[2] Hopefully, the standard Ubuntu packages (at least in source form)
will be available.

[3] ... and in the hope that they support standard GNU style symbol
visibility.

**Comment by tjhei** [comment]
> I do not consuder (2.) to be an option either. Cygwin, MinGW, MinGW-W64
> had the last time that I checked broken toolchains with ICE wrt name
> mangling and maximum section support in dynamic objects.

I am hoping that the clang toolchain might work better than gcc, but you might be right here.

> 5. Wait for the native Linux ABI for Windows 10

Ah, yes. Waiting is a convenient strategy. :stuck_out_tongue_winking_eye:

**Comment by bangerth** [comment]
I also don't see any convenient way. The manual approach is just too much work, and like @tamiko points out, it will lead to an endless set of patches where we have to add annotations we forgot in the original patch. Splitting things into multiple libraries will not work because they will have circular dependencies. That's why we put everything into one library back in the day, after struggling for a long time with three.

It may be that waiting for the MSVC-with-clang solution and a new file format may be the best we can do. In the meantime, static libraries are not great but we could at least occasionally use them to test. I don't see any other alternative, sadly.

**Comment by bangerth** [comment]
Out of curiosity, do we have any idea how many symbols we currently (try to) export?

**Comment by bangerth** [comment]
I'm asking because I could imagine that we can find *classes of symbols* that maybe we don't need to export. If we're close enough to the limit, maybe we can just remove those from the export list.

**Comment by bangerth** [comment]
Other links from a different project that struggled with this:
* https://ghc.haskell.org/trac/ghc/ticket/5987  (starting with comment 54)
* https://ghc.haskell.org/trac/ghc/wiki/DynamicGhcPrograms
* https://ghc.haskell.org/trac/ghc/wiki/WindowsDynamicLinking

**Comment by tjhei** [comment]
> Out of curiosity, do we have any idea how many symbols we currently (try to) export?

See above: "The debug lib is trying to export about 300'000 symbols".

**Comment by tjhei** [comment]
> static libraries are not great but we could at least occasionally use them to test

I can not run the whole testsuite (several TB in data, would take weeks). So I don't trust any of our code to work. Remember intel 13 compilers miscompiling ``FullMatrix``? Users trying to use windows are typically not experienced enough to deal with bugs inside compilers/deal.II internals. That means it is irresponsible to talk about "supporting windows". That is why I don't see this as an option. :-(

**Comment by bangerth** [comment]
Yes, I don't disagree. I was just thinking of spot checking some tests, but you're right that that just gives an overview of whether it works in general, not that a specific feature actually works.

Re 300,000: On linux, the .so has 187,704 symbol names. By symbol class, I get this:
```
/lib> for i in `nm libdeal_II.so.8.4.0-pre  | perl -p -e 's/^[0-9a-z]* *(.) .*/\1/g;' | sort | uniq` ; do echo $i `nm libdeal_II.so.8.4.0-pre | grep " $i " | wc -l` ; done
A 3
b 349
B 52
d 1686
D 211
r 16105
R 72
t 5250
T 3218
u 1051
U 1340
V 30589
w 19
W 127759
```
So not all of those are actually exports. On one of the pages I linked to from above, I saw that the default appears to be to re-export all symbols one imports from other libs, but that is not necessary if these others libs also appear on the linker line separately. I don't know how many of those 300,000 that would affect.

We could probably get rid of a lot of the weak symbols by disabling explicit instantiations and instead `#include`ing the `.template.h` file as we did recently for one file for MS VC anyway. I'm not sure that would get us below the limit.

**Comment by loylick** [comment]
"Placement and what you need to export is incompatible with the syntax expected by gcc/clang for *nix. "
GCC compiler perfectly understands dllexport syntax. I use it in my QT program working both under windows and under linux. You define MY_LIBRARY in the Makefile for the module with myClass code, in other modules this variable is undefined. 
#if defined(MY_LIBRARY)
#  define MY_LIBRARY_EXPORT dllexport
#else
#  define MY_LIBRARY_EXPORT dllimport
#endif

class MY_LIBRARY_EXPORT  myClass
{
}
You do not need to write __declspec( dllexport )  before every function.

**Comment by loylick** [comment]
Splitting things into multiple libraries will not work because they will have circular dependencies.  
Circular dependencies are compilable too. I did it several times using two stage compilation. First, one compiles libs with /FORCE flag which produce *.lib files with exported symbols in spite of missing dependencies. The second stage on builds dlls one by one, using this *.lib files to resolve missing dependencies.

**Comment by loylick** [comment]
> That would of course be enough. But it would require annotating thousands or tens of thousands of places :-(
That does not seem to be too much of a problem.  Once you make a list of classes to be exported you can run a script which would annotate all at once.

**Comment by bangerth** [comment]
It would be a long list. In fact I think it should include all classes declared in header files. 

Would you be interested in coming up with a trial patch based on such a script?

**Comment by loylick** [comment]
Do you mean that all classes with headers are used by end user? do not you have classes for internal use?

**Comment by tamiko** [comment]
@loylick 
> GCC compiler perfectly understands dllexport syntax. [...]

I wasn't terribly precise in what I meant.

There are subtle differences between `__declspec(dllexport)` and the visibility attributes set by `__attribute__((visibility("default")))`, etc.

The question is not so much about what happens for a simple showcase with a class, but how does symbol visibility actually look like for our templated library parts. You have to keep in mind that over the last 10 years a number of different strategies were used that lead to the fairly heterogeneous (beautiful :-) ) state of library we have.

So, in principle we can set up the buildsystem to define suitable import/export macros in a short amount of time and we can happily start to annotate header files. I simply fear that given the amount of highly complex and subtle corner cases we have, such as
 - non-templated code
 - templated code, only declaration in header
 - templated code, definition and declaration in header
intermixed, will result in subtle and unexpected visibility differences between platforms.

In order to reasonable be able to maintain this, we would have to restructure major parts of the library first. For example,
 - finish the transition to the declaration header / definition header / instantiation file triplet,
 - split up the library next.

Then, one could, for example, also think about providing a template library for windows with minimal dll, etc.

> Do you mean that all classes with headers are used by end user?

Pretty much yes. 

> Debug lib takes 1.4Gb I wonder how you managed to put so much stuff in it. In comparison PETSC takes 25MB, SLEPC 5Mb.

Because PETSc is in contrast non-templated C code. We, on the other hand, happily instantiate our templated code for a large number of use cases. The difference in size comparison are by the way not that drastic:
```
-rwxr-xr-x 1 root root 156050592 Sep  2 21:29 libdeal_II.g.so.9999
-rwxr-xr-x 1 root root 109769160 Sep  2 21:29 libdeal_II.so.9999
-rwxr-xr-x 1 root root  13505736 Aug 29 19:52 libpetsc.so.3.7.3
```
It's the debug symbols that make the large difference:
```
-rw-r--r-- 1 root root 1014599192 Sep  2 21:29 libdeal_II.g.so.9999.debug
-rw-r--r-- 1 root root  817735080 Sep  2 21:29 libdeal_II.so.9999.debug
-rw-r--r-- 1 root root   40395320 Aug 29 19:52 libpetsc.so.3.7.3.debug
```
Btw. for a fair comparison multiple the size for petsc by a suitable factor of different PETSC_SCALAR_TYPE instantiations (and independent libraries) you would have to compile.

Why we do that? Well, *nix simply let's us do that, that's why.

**Comment by jppelteret** [comment]
Just as a side note, according to one of the forum members deal.II seems to work quite well through Bash on Windows, available on Windows 10. More info in [this post](https://groups.google.com/forum/#!topic/dealii/fNZfonNFK8o).

**Comment by zhaofeng-shu33** [comment]
Splitting the library should resolve circular dependencies. In windows it can be solved with ".exp" file, see 
https://msdn.microsoft.com/en-us/library/kkt2hd12.aspx.  As https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/ points out,  if we use ".def" file to export symbols, "Handling class static data members " must be done to avoid unresolved external symbols about static class data.

**Comment by krishnakumarg1984** [comment]
The Windows Subsystem for Linux has undergone dramatic changes and rewrites in the recent versions of windows and now is based upon a native linux kernel embedded inside windows! This is called WSL2. Maybe this can make the windows install/use smoother for users. We may have to potentially update some documentation on this.

**Comment by drwells** [comment]
This issue tracks possible workarounds for MSVC, though; advancements in WSL improve the situation on Windows in a different way.

**Comment by krishnakumarg1984** [comment]
I see. Thank you for explaining. So far, I didn't see a native Windows installer for deal.II (although I personally do not care about this). I thought they mentioned WSL somewhere on the wiki or website, so I thought of letting the maintainers know that WSL has been rewritten from scratch to provide WSL2 (with a full linux kernel embedded inside along with the standard userland utilities).  Therefore, a lot of things should be easier, and thought that maybe we may need to update the wiki if necessary.

**Comment by krishnakumarg1984** [comment]
Has there been improvements in this front? Is a native MSVC compilable version of `deal.ii` available?

**Comment by masterleinad** [comment]
Yes, in fact, we have CI checks for `Visual Studio 16 2019` and `Visual Studio 15 2017`.

**Comment by krishnakumarg1984** [comment]
That's great news. However, a lot of solvers and libraries etc won't work on Windows, right? During compile time, do you just skip those dependences?

**Comment by masterleinad** [comment]
Yes, we check a basic configuration: https://github.com/dealii/dealii/blob/4cf8f26cbf26f12e630aff124cd112fb3f24180e/.github/workflows/windows.yml#L27.

**Comment by krishnakumarg1984** [comment]
Ok. Then maybe it is in a reasonable shape to provide a native installer for deal.ii on Windows machines in the project's downoad page https://dealii.org/download.html

**Comment by tjhei** [comment]
We compile cleanly on MSVC like @masterleinad says, but we have not made progress with dynamic linking.

Not a single of our main developers uses MSVC, so I am hesitant to advertise it more (limited performance, limited testing, limited external dependencies) or spend any effort to provide installers. 

Somewhat happier news: WSL is working very well for my students this semester and I would recommend that 100% over native windows.

**Comment by bangerth** [comment]
Just for the record, there are also a couple industrial projects that compile deal.II natively for Windows so these programs can interact seemlessly with other native Windows applications. (But I believe that all of these applications link statistically -- which is not all that much of an issue if all you care about is building and linking once.)
